firedrake/interpolation.py

```diff
@@ -645,3 +728,161 @@ def hash_expr(expr):
     domain_numbering = {d: i for i, d in enumerate(ufl.domain.extract_domains(expr))}
     coefficient_numbering = {c: i for i, c in enumerate(extract_coefficients(expr))}
     return compute_expression_signature(expr, {**domain_numbering, **coefficient_numbering})
+
+
+class VomOntoVomCallable(object):
+    """Callable that maps from one ``VertexOnlyMesh`` to it's intput ordering
+    ``VertexOnlyMesh``, or vice versa.
+
+    Parameters
+    ----------
+    V : `.FunctionSpace`
+        The P0DG function space (which may be vector or tensor valued) on the
+        source vertex-only mesh.
+    source_vom : `.VertexOnlyMesh`
+        The vertex-only mesh we interpolate from.
+    target_vom : `.VertexOnlyMesh`
+        The vertex-only mesh we interpolate to.
+    expr : `ufl.Expr`
+        The expression to interpolate. If ``arguments`` is not empty, those
+        arguments must be present within it.
+    arguments : list of `ufl.Argument`
+        The arguments in the expression. These are not extracted from expr here
+        since, where we use this, we already have them.
+    """
+
+    def __init__(self, V, source_vom, target_vom, expr, arguments):
+        source_vom = source_vom
+        reduce = False
+        broadcast = False
+        try:
+            if source_vom.input_ordering is target_vom:
+                reduce = True
+                original_vom = source_vom
+        except AttributeError:
+            pass
+        try:
+            if target_vom.input_ordering is source_vom:
+                broadcast = True
+                original_vom = target_vom
+        except AttributeError:
+            pass
+        if not (reduce or broadcast):
+            raise ValueError(
+                "The target vom and source vom must be linked by input ordering!"
+            )
```

Comment:
Maybe just set `input_ordering` (or whatever name you decide to use) ` = None` and remove these `try`/`except`?